nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,159 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(
+        {%{shape: {m, k} = shape_q} = q_holder,
+         %{shape: {k, n} = shape_r, type: {_, num_bits} = type} = r_holder},
+        tensor,
+        opts
+      ) do
+    mode = opts[:mode]
+
+    r =
+      if mode == :reduced do
+        tensor[[0..(k - 1), 0..(n - 1)]]
+      else
+        tensor
+      end
+      |> Nx.as_type(type)
+      |> to_binary()
+
+    shape_h = {k, k}
+
+    {q_bin, r_bin} =
+      for i <- 0..(n - 1), reduce: {nil, r} do
+        {q, r} ->
+          # a = r[[i..(k - 1), i]]
+          {a, num_rows_a, _, _} =
+            for <<x::bitstring-size(num_bits) <- r>>, reduce: {<<>>, 0, 0, 0} do
+              {data, num_rows_a_prev, row, col} ->
+                {item, num_rows_a} =
+                  if row >= i and col == i do
+                    {x, num_rows_a_prev + 1}
+                  else
+                    {"", num_rows_a_prev}
+                  end
+
+                {row, col} =
+                  if col + 1 == n do
+                    {row + 1, 0}
+                  else
+                    {row, col + 1}
+                  end
+
+                {data <> item, num_rows_a, row, col}
+            end
+
+          h_bin = householder_reflector(a, num_rows_a, type, k)
+
+          # If we haven't allocated Q yet, let Q = H1
+          q =
+            if is_nil(q) do
+              padding_length = (m - k) * k
+
+              zero = number_to_binary(0, type)
+
+              h_bin <> String.duplicate(zero, padding_length)
+            else
+              dot_binary(q, shape_q, h_bin, shape_h, type)
+            end
+
+          r = dot_binary(h_bin, shape_q, r, shape_r, type)
+
+          {q, r}
+      end
+
+    q = from_binary(q_holder, q_bin)
+    r = from_binary(r_holder, r_bin)
+
+    case mode do
+      :complete ->
+        {q, r}
+
+      :reduced ->
+        {q, r[[0..(k - 1), 0..(n - 1)]]}
+    end
+  end
+
+  defp dot_binary(b1, shape1, b2, shape2, {_, s} = type) do
+    # matrix multiplication for when b1 and b2 are matrices
+    # represented in binary form
+    fun = fn lhs, rhs, acc ->
+      res = binary_to_number(lhs, type) * binary_to_number(rhs, type) + acc
+      {res, res}
+    end
+
+    v1 = aggregate_axes(b1, [1], shape1, s)
+    v2 = aggregate_axes(b2, [0], shape2, s)
+
+    for b1 <- v1, b2 <- v2, into: <<>> do
+      {bin, _acc} = bin_zip_reduce_axis(b1, b2, s, s, <<>>, 0, fun)
+      scalar_to_binary(bin, type)
+    end
+  end
+
+  defp householder_reflector(a_bin, num_rows_a, {_, num_bits} = type, target_k) do
+    # We receive the target shape so the binary is already pre-generated with 1s in the
+    # main diagonal for rows < k - n
+    <<head::bitstring-size(num_bits), tail::bitstring>> = a_bin
+
+    norm_a_squared =
+      for <<x_bin::bitstring-size(num_bits) <- a_bin>>, reduce: 0 do
+        acc ->
+          x = binary_to_number(x_bin, type)
+          acc + x * x
+      end
```

Comment:
Generally speaking, we want to discourage using `binary_to_number`. It is more efficient to use the match_types syntax:

```suggestion
      match_types [input_type] do
        for <<match!(x, 0) <- a_bin>>, reduce: 0 do
          acc ->
            x = read!(x, 0)
            acc + x * x
        end
      end
```

Reply:
Or better yet, convert `a_bin` to `a_list` at the beginning of the function with this:

    for <<match!(x, 0) <- a_bin>>, do: read!(x, 0)

And then keep a_bin and v_bin as lists. I think it will simplify and speed up the rest of the code considerably.